src/Stratis.Features.FederatedPeg/InputConsolidation/InputConsolidator.cs

```diff
@@ -0,0 +1,240 @@
+﻿using System;
+using System.Collections.Generic;
+using System.Linq;
+using System.Threading.Tasks;
+using Microsoft.Extensions.Logging;
+using NBitcoin;
+using Stratis.Bitcoin.Features.Wallet.Interfaces;
+using Stratis.Bitcoin.Primitives;
+using Stratis.Bitcoin.Signals;
+using Stratis.Features.FederatedPeg.Events;
+using Stratis.Features.FederatedPeg.Interfaces;
+using Stratis.Features.FederatedPeg.Payloads;
+using Stratis.Features.FederatedPeg.TargetChain;
+using Stratis.Features.FederatedPeg.Wallet;
+
+namespace Stratis.Features.FederatedPeg.InputConsolidation
+{
+    public class InputConsolidator : IInputConsolidator
+    {
+        public static readonly Money ConsolidationFee = Money.Coins(0.005m); // 50 inputs. This is roughly half the same as fee on withdrawals
+
+        private readonly IFederatedPegBroadcaster federatedPegBroadcaster;
+        private readonly IFederationWalletTransactionHandler transactionHandler;
+        private readonly IFederationWalletManager walletManager;
+        private readonly IBroadcasterManager broadcasterManager;
+        private readonly IFederatedPegSettings settings;
+        private readonly ILogger logger;
+        private readonly Network network;
+
+        private bool signingInProgress;
+
+        private bool fullySigned;
+
+        /// <summary>
+        /// Used to ensure only one operation is happening at a time.
+        /// </summary>
+        private readonly object lockObj = new object();
+
+        /// <inheritdoc />
+        public Transaction PartialTransaction { get; private set; }
+
+        public InputConsolidator(IFederatedPegBroadcaster federatedPegBroadcaster,
+            IFederationWalletTransactionHandler transactionHandler,
+            IFederationWalletManager walletManager,
+            IBroadcasterManager broadcasterManager,
+            IFederatedPegSettings settings,
+            ILoggerFactory loggerFactory,
+            ISignals signals,
+            Network network)
+        {
+            this.federatedPegBroadcaster = federatedPegBroadcaster;
+            this.transactionHandler = transactionHandler;
+            this.walletManager = walletManager;
+            this.broadcasterManager = broadcasterManager;
+            this.network = network;
+            this.settings = settings;
+            this.logger = loggerFactory.CreateLogger(this.GetType().FullName);
+            signals.Subscribe<WalletNeedsConsolidation>(this.StartConsolidation);
+        }
+
+        /// <inheritdoc />
+        public void StartConsolidation(WalletNeedsConsolidation trigger)
+        {
+            // TODO: This is feral is there a better way
```

Comment:
Perhaps don't call this directly via the signal mechanism. Create a `WaitLoop` including this code:

```
Console.WriteLine("Waiting for hardware thread to do it's work...");
signal.WaitOne(); // block thread until we are signaled.
signal.Dispose();
Console.WriteLine("Got our signal! we can continue.");
```

Call `signal.Set()` from a new subscribing method.